lms/djangoapps/grades/course_grade_factory.py

```diff
@@ -160,13 +161,16 @@ def _read(user, course_data):
         )
 
     @staticmethod
-    def _update(user, course_data, force_update_subsections=False):
+    def _update(user, course_data, force_update_subsections=False, send_grade_signals=True):
         """
-        Computes, saves, and returns a CourseGrade object for the
-        given user and course.
-        Sends a COURSE_GRADE_CHANGED signal to listeners and
-        COURSE_GRADE_NOW_PASSED if learner has passed course or
-        COURSE_GRADE_NOW_FAILED if learner is now failing course
+        Computes, saves, and returns a CourseGrade object for the given user and course.
+
+        send_grade_signals defines if signals should be sent. Use it to avoid recursion issues in
+        cases when the signal listener trying to get grades but Persistent Grades are disabled.
+        If True - sends:
+            COURSE_GRADE_CHANGED signal to listeners and
+            COURSE_GRADE_NOW_PASSED if learner has passed course or
+            COURSE_GRADE_NOW_FAILED if learner is now failing course
```

Comment:
This is probably any place where the `CourseGradeFactory().read()` used (because it calls the `_update()` if there is no persistent grade present).

Some examples of the signal receivers:
- evaluation of any milestone relationships which are attached
    to the course grade;
- Notifies the Credentials IDA about certain grades it needs for its records, when a grade changes;
- update minimum grade requirement status;
- generate a certificate task;
- etc

Reply:
> evaluation of any milestone relationships which are attached
> to the course grade;
> Notifies the Credentials IDA about certain grades it needs for its records, when a grade changes;
> update minimum grade requirement status;
> **generate a certificate task;**
> etc

But the whole is to not send the course grade related signals when generating the certificate task right?

Reply:
Yes, it's the main goal to break the recursive grade recalculation when the certificate is generated